File: test/arguments.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

require 'msgpack/rpc'
require 'test/unit'
require './dataset'
require './const'

class SyncCall < Test::Unit::TestCase

  def get_session
    session = MessagePack::RPC::Client.new(HOSTNAME, PORT)
    session.timeout = TIMEOUT
    session
  end

  def test_arguments
    dataset = DataSet.new.dataset_for_arguments
    session = get_session
    dataset.each do |d|
      p d
      result = session.call(d["method"], d["arg1"], d["arg2"], d["arg3"], d["arg4"])
      p result
      assert_equal(result[0], d["expect"][0])
      assert_equal(result[1], d["expect"][1])
      assert_equal(result[2], d["expect"][2])
      assert_equal(result[3], d["expect"][3])
    end
    session.close()
  end
end

File: test/arguments.jl
#!/usr/bin/env julia

include("../src/MsgPackRpcClient.jl")
include("const.jl")
include("dataset.jl")
using MsgPackRpcClient
using Base.Test

function get_max_arg(data; prefix = "arg")
  i = 1
  while i < 65536
    if haskey(data, "$(prefix)$(i)")
      i += 1
    else
      break
    end
  end
  i - 1
end

function test_data(data, number = 1)
  for d in data
    println("$(number): $d ")
    #try
      max_arg = get_max_arg(d)

      if haskey(d, "arg4")
        result = MsgPackRpcClient.call(session, d["method"], d["arg1"], d["arg2"], d["arg3"], d["arg4"]; sync = true)
      elseif haskey(d, "arg3")
        result = MsgPackRpcClient.call(session, d["method"], d["arg1"], d["arg2"], d["arg3"]; sync = true)
      elseif haskey(d, "arg2")
        result = MsgPackRpcClient.call(session, d["method"], d["arg1"], d["arg2"]; sync = true)
      elseif haskey(d, "arg1")
        result = MsgPackRpcClient.call(session, d["method"], d["arg1"]; sync = true)
      else
        error
      end

      #println("result = ", result)
      #println("first(result) = ", first(result))
      if haskey(d, "expect")
        for i = 1:max_arg
          @test result[i] == d["expect"][i]
        end
      end
      println("  => OK")
    #catch
    #  #println("Error on d[\"arg\"] = ", d["arg"])
    #  println("Error")
    #end
    number += 1
  end
end

session = MsgPackRpcClientSession.create()

test_data(get_dataset_for_arguments())

session.destroy()

